src/include/plan_node/abstract_plan_node.h

```diff
@@ -0,0 +1,141 @@
+#pragma once
+
+#include <cstdint>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "common/hash_util.h"
+#include "common/json.h"
+#include "plan_node/output_schema.h"
+#include "plan_node/plan_node_defs.h"
+
+// TODO(Gus,Wen): Add equaility operator and hash function support for output_schema
+
+namespace terrier::plan_node {
+//===--------------------------------------------------------------------===//
+// Abstract Plan Node
+//===--------------------------------------------------------------------===//
+
+class AbstractPlanNode {
+ public:
+  explicit AbstractPlanNode(std::shared_ptr<OutputSchema> output_schema) : output_schema_(std::move(output_schema)) {}
+
+  // For Deserialization and DDL statements
+  AbstractPlanNode() = default;
+
+  virtual ~AbstractPlanNode() = default;
+
+  //===--------------------------------------------------------------------===//
+  // Children Helpers
+  //===--------------------------------------------------------------------===//
+
+  void AddChild(std::unique_ptr<AbstractPlanNode> &&child) { children_.emplace_back(std::move(child)); }
```

Comment:
Didn't we say this class is going to be immutable?

Reply:
`AddChild` is only used when constructing the plan node. Do you think it would be better to pass in the child plans into the constructor? That should make them immutable.

Reply:
Marking as resolved, we're changing the plan nodes to use a builder pattern.